strawberry/schema/execute_context.py

```diff
@@ -0,0 +1,127 @@
+from typing import Any, Dict, List, Optional, Union
+
+from promise import Promise, is_thenable, promise_for_dict
+
+from graphql import (
+    ExecutionContext,
+    GraphQLError,
+    GraphQLField,
+    GraphQLFieldResolver,
+    GraphQLObjectType,
+    GraphQLOutputType,
+    GraphQLResolveInfo,
+)
+from graphql.execution.values import get_argument_values
+from graphql.language import FieldNode
+from graphql.pyutils import AwaitableOrValue, Path, Undefined
+
+
+class ExecutionContextWithPromise(ExecutionContext):
+    def build_response(self, data):
+        if is_thenable(data):
+            original_build_response = super().build_response
+
+            def on_rejected(error):
+                self.errors.append(error)
+                return None
+
+            def on_resolve(data):
+                return original_build_response(data)
+
+            promise = data.catch(on_rejected).then(on_resolve)
+            return promise
+        return super().build_response(data)
+
+    def resolve_field_value_or_error(
+        self,
+        field_def: GraphQLField,
+        field_nodes: List[FieldNode],
+        resolve_fn: GraphQLFieldResolver,
+        source: Any,
+        info: GraphQLResolveInfo,
+    ) -> Union[Exception, Any]:
+        """Resolve field to a value or an error.
+
+        Isolates the "ReturnOrAbrupt" behavior to not de-opt the resolve_field()
+        method. Returns the result of resolveFn or the abrupt-return Error object.
+
+        For internal use only.
+        """
+        try:
+            # Build a dictionary of arguments from the field.arguments AST, using the
+            # variables scope to fulfill any variable references.
+            args = get_argument_values(field_def, field_nodes[0], self.variable_values)
+
+            # Note that contrary to the JavaScript implementation, we pass the context
+            # value as part of the resolve info.
+            result = resolve_fn(source, info, **args)
+            if is_thenable(result):
```

Comment:
What's the purpose of this if? Wouldn't the program flow be the same without it?

Reply:
Oh yeah I don't need this anymore. In the original code there was an `if is_awaitable` branch as well which I removed.